### PR TITLE
Make list API consistent

### PR DESCRIPTION
**Pull request by MikaelJohnssonModelon**
Purpose:
Make the API more consistent and readable API. Get rid of a undesired dependency.

Problems fixed by this PR:

1. Inconsistent naming for list types
2. Preferred API structure:

> fmi3_import_XX_list_t* fmi3_import_get_XX_list(fmi3_import_t* fmu);
> size_t                 fmi3_import_get_XX_list_size(fmi3_import_XX_list_t* vl)
> fmi3_import_XX_t*      fmi3_import_get_XX(fmi3_import_XX_list_t* vl, size_t index);

3. Do not have any xml source header in the import headers.

These issues are fixed by: 
Rename a number of list API methods to be more consistent. 
Prefer _list_t and _list_size as the standard over other alternatives.
And other small fixes.
Copied the needed parts to a new import header

### CODE REVIEW

File: src/Import/src/FMI3/fmi3_import_variable.c [hunk 1]
```diff
@@ -580,15 +580,15 @@ int fmi3_import_get_enum_variable_max(fmi3_import_enum_variable_t* v) {
 // Alias variable
 // -----------------------------------------------------------------------------
 
-fmi3_import_alias_variables_t* fmi3_import_get_variable_aliases(fmi3_import_variable_t* v) {
+fmi3_import_alias_variable_list_t* fmi3_import_get_variable_aliases(fmi3_import_variable_t* v) {
     return fmi3_xml_get_variable_aliases(v);
 }
 
-size_t fmi3_import_get_alias_variables_number(fmi3_import_alias_variables_t* aliases) {
+size_t fmi3_import_get_alias_variables_number(fmi3_import_alias_variable_list_t* aliases) {
```
Comment (PeterMeisrimelModelon) [inline src/Import/src/FMI3/fmi3_import_variable.c:587]: Rename function to `fmi3_import_get_alias_list_size` ?

File: src/XML/include/FMI3/fmi3_xml_type.h [hunk 2]
```diff
@@ -35,13 +35,13 @@ extern "C" {
     @{
     */
 
-fmi3_xml_type_definitions_t* fmi3_xml_get_type_definitions(fmi3_xml_model_description_t* md);
+fmi3_xml_type_definition_list_t* fmi3_xml_get_type_definitions(fmi3_xml_model_description_t* md);
 
 /* Convert base type constant to string */
 
-unsigned int fmi3_xml_get_type_definition_number(fmi3_xml_type_definitions_t* td);
+unsigned int fmi3_xml_get_type_definition_number(fmi3_xml_type_definition_list_t* td);
```
Comment (PeterMeisrimelModelon) [inline src/XML/include/FMI3/fmi3_xml_type.h:42]: `size_t fmi3_xml_get_type_definition_list_size` ?

File: src/XML/include/FMI3/fmi3_xml_type.h [hunk 3]
```diff
@@ -95,9 +95,9 @@ fmi3_uint8_t fmi3_xml_get_uint8_type_max(fmi3_xml_int_typedef_t*);
 int fmi3_xml_get_enum_type_min(fmi3_xml_enumeration_typedef_t*);
 int fmi3_xml_get_enum_type_max(fmi3_xml_enumeration_typedef_t*);
 unsigned int  fmi3_xml_get_enum_type_size(fmi3_xml_enumeration_typedef_t*);
-const char* fmi3_xml_get_enum_type_item_name(fmi3_xml_enumeration_typedef_t*, unsigned int  item);
```
Comment (PeterMeisrimelModelon) [inline src/XML/include/FMI3/fmi3_xml_type.h:98]: See comments in corresponding .c file

File: src/XML/include/FMI3/fmi3_xml_variable.h [hunk 4]
```diff
@@ -48,9 +48,9 @@ fmi3_value_reference_t fmi3_xml_get_variable_vr(fmi3_xml_variable_t*);
 fmi3_xml_variable_typedef_t* fmi3_xml_get_variable_declared_type(fmi3_xml_variable_t*);
 fmi3_base_type_enu_t fmi3_xml_get_variable_base_type(fmi3_xml_variable_t*);
 
-fmi3_xml_alias_variables_t* fmi3_xml_get_variable_aliases(fmi3_xml_variable_t* v);
```
Comment (PeterMeisrimelModelon) [inline src/XML/include/FMI3/fmi3_xml_variable.h:51]: See comments in corresponding .c file

File: src/Import/include/FMI3/fmi3_import_dimension_list.h [hunk 5]
```diff
@@ -36,23 +36,17 @@ extern "C" {
...
  */
 
-/** \brief Allocate an empty list (internal only) */
-fmi3_import_dimension_list_t* fmi3_import_alloc_dimension_list(fmi3_import_t* fmu, fmi3_import_variable_t* v);
-
-/**  \brief Free a dimension list. Note that dimension lists are allocated dynamically and must be freed when not needed any longer 
-    @param dl A dimension list.
-*/
-FMILIB_EXPORT void fmi3_import_free_dimension_list(fmi3_import_dimension_list_t* dl);
+/** \brief Get a list of dimensions from the import variables */
```
Comment (PeterMeisrimelModelon) [inline src/Import/include/FMI3/fmi3_import_dimension_list.h:42]:
```suggestion
/** \brief Get a list of dimensions for an import variable */
```

File: src/Import/include/FMI3/fmi3_import_type.h [hunk 6]
```diff
@@ -49,16 +49,16 @@ typedef struct fmi3_xml_clock_typedef_t fmi3_import_clock_typedef_t;
 /** \brief Opaque general variable type definition object. */
 typedef struct fmi3_xml_variable_typedef_t fmi3_import_variable_typedef_t;
 /** \brief Opaque list of the type definitions in the model */
-typedef struct fmi3_xml_type_definitions_t fmi3_import_type_definitions_t;
+typedef struct fmi3_xml_type_definition_list_t fmi3_import_type_definitions_t;
 /**@} */
 
 /** \brief Get the number of available type definitions */
-FMILIB_EXPORT unsigned int fmi3_import_get_type_definition_number(fmi3_import_type_definitions_t* td);
+FMILIB_EXPORT unsigned int fmi3_import_get_type_definition_list_size(fmi3_import_type_definitions_t* td);
```
Comment (PeterMeisrimelModelon) [inline src/Import/include/FMI3/fmi3_import_type.h:56]: unsigned int -> size_t?

**Review by PeterMeisrimelModelon — CHANGES_REQUESTED** [review]
I only reviewed the API consistency changes for now and not the include changes, the proposed changes look quite involved and I think we are currently lacking proper testing to diagnose any issues there in the first place.

Looks good overall. Left a bunch of comments, since a few cases were missed.

Along with unifying the return type of _get_list_size functions (I'd propose size_t over unsigned int), I think we should also unify the corresponding _get_ functions that take a list + index to then also use size_t as the index type.

Also, when rebasing to master, there will likely be a whole lot of conflict or changes that need to be migrated, since fmi3_import_arrays_test.c was converted to .cpp.

File: src/Import/include/FMI3/fmi3_import_type.h [hunk 7]
```diff
@@ -49,16 +49,16 @@ typedef struct fmi3_xml_clock_typedef_t fmi3_import_clock_typedef_t;
 /** \brief Opaque general variable type definition object. */
 typedef struct fmi3_xml_variable_typedef_t fmi3_import_variable_typedef_t;
 /** \brief Opaque list of the type definitions in the model */
-typedef struct fmi3_xml_type_definitions_t fmi3_import_type_definitions_t;
```
[the remark this answers is not in the document]
Comment (MikaelJohnssonModelon) [reply src/Import/include/FMI3/fmi3_import_type.h:52]: Yes, also found a few others I missed.

File: src/Import/src/FMI3/fmi3_import.c [hunk 8]
```diff
@@ -274,13 +274,13 @@ double fmi3_import_get_default_experiment_step(fmi3_import_t* fmu) {
     return fmi3_xml_get_default_experiment_step(fmu->md);
 }
 
-fmi3_import_unit_definitions_t* fmi3_import_get_unit_definitions(fmi3_import_t* fmu) {
```
[the remark this answers is not in the document]
Comment (MikaelJohnssonModelon) [reply src/Import/src/FMI3/fmi3_import.c:277]: Fixed

File: src/Import/src/FMI3/fmi3_import.c [hunk 9]
```diff
@@ -274,13 +274,13 @@ double fmi3_import_get_default_experiment_step(fmi3_import_t* fmu) {
...
+fmi3_import_unit_definition_list_t* fmi3_import_get_unit_definitions(fmi3_import_t* fmu) {
     if (!fmi3_import_check_has_FMU(fmu)) return NULL;
 
     return fmi3_xml_get_unit_definitions(fmu->md);
 }
 
-unsigned int  fmi3_import_get_unit_definitions_number(fmi3_import_unit_definitions_t* ud) {
```
[the remark this answers is not in the document]
Comment (MikaelJohnssonModelon) [reply src/Import/src/FMI3/fmi3_import.c:283]: Yes, fixed.

File: src/Import/src/FMI3/fmi3_import_type.c [hunk 10]
```diff
@@ -19,7 +19,7 @@
 
 #include "fmi3_import_impl.h"
 
-unsigned int fmi3_import_get_type_definition_number(fmi3_import_type_definitions_t* td) {
```
[the remark this answers is not in the document]
Comment (MikaelJohnssonModelon) [reply src/Import/src/FMI3/fmi3_import_type.c:22]: Yes and corrected the name of the method too. To a "_list_size" method.

File: src/Import/src/FMI3/fmi3_import_type.c [hunk 11]
```diff
@@ -240,16 +240,16 @@ unsigned int  fmi3_import_get_enum_type_size(fmi3_import_enumeration_typedef_t*
     return fmi3_xml_get_enum_type_size(t);
 }
 
-const char* fmi3_import_get_enum_type_item_name(fmi3_import_enumeration_typedef_t* t, unsigned int  item) {
```
[the remark this answers is not in the document]
Comment (MikaelJohnssonModelon) [reply src/Import/src/FMI3/fmi3_import_type.c:243]: Ok, I though this was some enum types and enum literal kinda situation in which case enum_type and enum_type_item sounds kinda clunky.

File: Test/FMI3/fmi3_import_arrays_test.c [hunk 12]
```diff
@@ -233,7 +233,7 @@ static int test_array_ok_32(fmi3_import_t* xml, char* varName, fmi3_float32_t* s
     is_array = fmi3_import_variable_is_array(v);
     ASSERT_MSG(is_array, "wrong variable type: expected array, but wasn't");
 
-    dimList = fmi3_import_get_variable_dimension_list(xml, v); /* allocates memory */
+    dimList = fmi3_import_get_variable_dimension_list(v); /* allocates memory */
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply Test/FMI3/fmi3_import_arrays_test.c:236]: No they are now wrong, feel free to fix

File: src/Import/src/FMI3/fmi3_import_type.h [hunk 13]
```diff
@@ -0,0 +1,311 @@
+/*
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/Import/src/FMI3/fmi3_import_type.h:1]: Yes I have no idea how that happened. Should be no changes made to it. Just remove

**Review by PeterMeisrimelModelon — APPROVED** [review]

**Pull request merged**